exonum/src/runtime/rust/mod.rs

```diff
@@ -28,6 +28,108 @@
+//! use exonum_crypto::Hash;
+//!
+//! // Determine the types of data that will be used in service transactions.
+//!
+//! #[derive(Debug, PartialEq, ProtobufConvert, BinaryValue, ObjectHash)]
+//! #[protobuf_convert(source = "doc_tests::Point")]
+//! pub struct Point {
+//!     pub x: i32,
+//!     pub y: i32,
+//! }
+//!
+//! #[derive(Debug, PartialEq, ProtobufConvert, BinaryValue, ObjectHash)]
+//! #[protobuf_convert(source = "doc_tests::CreateWallet")]
+//! pub struct CreateWallet {
+//!     pub name: String,
+//! }
+//!
+//! // You may create service-specific error types.
+//!
+//! #[derive(Debug, IntoExecutionError)]
+//! pub enum Error {
+//!     PointAlreadyExists = 0,
+//!     WalletAlreadyExists = 1,
+//! }
+//!
+//! // Define the transactions interface for your service by creating a trait with
+//! // the following attribute and method signatures.
+//! #[exonum_interface]
+//! pub trait Transactions {
+//!     // You may use `ExecutionError` directly.
+//!     fn create_wallet(
+//!         &self,
+//!         context: CallContext<'_>,
+//!         arg: CreateWallet,
+//!     ) -> Result<(), ExecutionError>;
+//!     // Also you can use for error any type which implements `Into<ExecutionError>`.
+//!     fn add_point(
+//!         &self,
+//!         context: CallContext<'_>,
+//!         arg: Point,
+//!     ) -> Result<(), Error>;
+//! }
+//!
+//! // In order for a service to process transactions, you have to implement the
+//! // `ServiceDispatcher` trait, which can be derived using the corresponding macro.
+//! // To explain runtime how to create instances of this service you have to implement
+//! // `ServiceFactory` trait by using the `ServiceFactory` derive macro.
+//! #[derive(Debug, ServiceDispatcher, ServiceFactory)]
+//! // Declare that a service implements a `Transactions` interface that was presented above.
+//! #[service_dispatcher(implements("Transactions"))]
+//! // By default macro uses crate name and version to provide artifact ID for this
+//! // service factory. You should only provide path to the generated Protobuf schema.
+//! #[service_factory(proto_sources = "exonum::proto::schema")]
+//! pub struct PointService;
+//!
+//! // Don't forget to implement the `Transactions` and `Service` traits for our service.
+//!
+//! impl Transactions for PointService {
+//!     fn create_wallet(
+//!         &self,
+//!         _context: CallContext<'_>,
+//!         _arg: CreateWallet,
+//!     ) -> Result<(), ExecutionError> {
+//!         // Some business logic...
+//!         Ok(())
+//!     }
+//!
+//!     fn add_point(
+//!         &self,
+//!         _context: CallContext<'_>,
+//!         _arg: Point
+//!     ) -> Result<(), Error> {
+//!         // Some business logic...
+//!         Ok(())
+//!     }
+//! }
+//!
+//! impl Service for PointService {
+//!     fn state_hash(&self, _data: BlockchainData<&dyn Snapshot>) -> Vec<Hash> {
+//!         Vec::new()
+//!     }
+//! }
+//! ```
```

Comment:
Maybe add also links to `cryptocurrency` and `cryptocurrency-advanced` to see examples closer to the real world?

Reply:
It will not work with `docs.rs`.